lib/Sema/CSSimplify.cpp

```diff
@@ -3925,6 +3930,42 @@ ConstraintSystem::simplifyDynamicTypeOfConstraint(
   return SolutionKind::Error;
 }
 
+ConstraintSystem::SolutionKind
+ConstraintSystem::simplifyOpaqueUnderlyingTypeConstraint(Type type1, Type type2,
+                                             TypeMatchOptions flags,
+                                             ConstraintLocatorBuilder locator) {
+  // Open the second type, which must be an opaque archetype, to try to
+  // infer the first type using its constraints.
+  auto opaque2 = type2->castTo<OpaqueTypeArchetypeType>();
+
+  // Open the generic signature of the opaque decl, and bind the "outer" generic
+  // params to our context. The remaining axes of freedom on the type variable
+  // corresponding to the underlying type should be the constraints on the
+  // underlying return type.
+  auto opaqueDecl = opaque2->getOpaqueDecl();
+  auto interfaceSignature = opaqueDecl->getOpaqueInterfaceGenericSignature();
+  OpenedTypeMap replacements;
+  openGeneric(nullptr, DC, interfaceSignature,
+              /*skip self*/ false,
+              locator, replacements);
+  
+  auto underlyingTyVar = openType(opaqueDecl->getUnderlyingInterfaceType(),
+                                  replacements);
+  assert(underlyingTyVar);
+  
+  if (auto dcSig = DC->getGenericSignatureOfContext()) {
+    for (auto param : dcSig->getGenericParams()) {
+      addConstraint(ConstraintKind::Bind,
+                    openType(param, replacements),
+                    DC->mapTypeIntoContext(param),
+                    locator);
+    }
+  }
+
+  addConstraint(ConstraintKind::Equal, type1, underlyingTyVar, locator);
```

Comment:
Just for my better understanding, could you please elaborate if this constraint has to be equality, or it could be conversion instead? It seems like in the non-opaque case solver creates a conversion constraint from contextual result to what we infer for `return`, but in this case I _think_ equality makes more sense because base is going to be a type variable with a set of requirements, is that correct? If so, I wonder if we should just bind these types instead of to pull requirements to the left type?

Reply:
bq. Are there cases where a Conversion constraint between type variables might still allow a more general set of types to be bound?

I can only think of implicit conversions being a difference between equals and conversion in that regard.

bq. Bind seems too specific since we still want to allow lvalues in the return statement to be loaded too.

Actually if you look at the `matchTypes` - both equal and bind share the same path and most likely going to bind types or merge equivalence classes of the type variables if either is a type variable.